include/proj/coordinatesystem.hh

```diff
@@ -0,0 +1,302 @@
+
+  private:
+    explicit AxisDirection(const std::string &nameIn);
+
+    static std::map<std::string, const AxisDirection *> axisDirectionRegistry;
+    static std::set<std::string> axisDirectionKeys;
+};
+
+// ---------------------------------------------------------------------------
+
+class Meridian;
+using MeridianPtr = std::shared_ptr<Meridian>;
+
+/** For WKT2. Used by CoordinateSystemAxis optionaly */
+class Meridian : public common::IdentifiedObject, public io::IWKTExportable {
+  public:
+    PROJ_DLL Meridian(const Meridian &other);
+    PROJ_DLL Meridian &operator=(const Meridian &other) = delete;
+    PROJ_DLL ~Meridian();
+
+    PROJ_DLL const common::Angle &longitude() const;
+
+    // non-standard
+    PROJ_DLL static MeridianPtr create(const util::PropertyMap &properties,
+                                       const common::Angle &longitude);
+
+    PROJ_DLL std::string exportToWKT(io::WKTFormatterNNPtr formatter)
+        const override; // throw(io::FormattingException)
+
+  protected:
+#ifdef DOXYGEN_ENABLED
+    Angle angle_;
+#endif
+
+    Meridian();
+    INLINED_MAKE_SHARED
+
+  private:
+    PROJ_OPAQUE_PRIVATE_DATA
+};
+
+// ---------------------------------------------------------------------------
+
+class CoordinateSystemAxis;
+using CoordinateSystemAxisPtr = std::shared_ptr<CoordinateSystemAxis>;
+using CoordinateSystemAxisNNPtr = util::nn<CoordinateSystemAxisPtr>;
+
+class CoordinateSystemAxis : public common::IdentifiedObject,
+                             public io::IWKTExportable {
+  public:
+    PROJ_DLL CoordinateSystemAxis(const CoordinateSystemAxis &other);
+    PROJ_DLL CoordinateSystemAxis &
+    operator=(const CoordinateSystemAxis &other) = delete;
+    PROJ_DLL ~CoordinateSystemAxis();
+
+    PROJ_DLL const std::string &axisAbbrev() const;
+    PROJ_DLL const AxisDirection &axisDirection() const;
+    PROJ_DLL const common::UnitOfMeasure &axisUnitID() const;
+    PROJ_DLL const util::optional<double> &minimumValue() const;
+    PROJ_DLL const util::optional<double> &maximumValue() const;
+    PROJ_DLL const MeridianPtr &meridian() const;
+
+    // Non-standard
+    PROJ_DLL static CoordinateSystemAxisNNPtr
+    create(const util::PropertyMap &properties, const std::string &abbreviation,
+           const AxisDirection &direction, const common::UnitOfMeasure &unit,
+           const MeridianPtr &meridian = nullptr);
+
+    PROJ_DLL std::string exportToWKT(io::WKTFormatterNNPtr formatter)
+        const override; // throw(io::FormattingException)
+    PROJ_DLL std::string exportToWKT(io::WKTFormatterNNPtr formatter,
+                                     int order) const;
+
+  private:
+    PROJ_OPAQUE_PRIVATE_DATA
+
+    CoordinateSystemAxis();
+    INLINED_MAKE_SHARED
+};
+
+// ---------------------------------------------------------------------------
+
+class CoordinateSystem : public common::IdentifiedObject,
+                         public io::IWKTExportable {
```

Comment:
Dumb question: can `CoordinateSystem` be something-else-exportable?

Reply:
At some point I guess there will be a IPROJStringExportable interface for things that can be converted to PROJ strings, like CRS or Conversion, but a CoordinateSystem itself probably not